File: cpp/stack/largest_rectangle.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define pb push_back
#define all(x) x.begin(), x.end()
#define endl '\n'

int largestRectangleArea(vector<int>& heights) {
    int n = heights.size();
    vector<int> foward(n, -1);
    vector<int> backwards(n, -1);
    stack<pair<int, int>> st;
    for (int i = 0; i < n; i++) {
        while (!st.empty() && heights[i] < st.top().first) {
            foward[st.top().second] = i - st.top().second - 1;
            st.pop();
        }
        st.push({ heights[i], i });
    }
    while (!st.empty()) st.pop();
    for (int i = n - 1; i >= 0; i--) {
        while (!st.empty() && heights[i] < st.top().first) {
            backwards[st.top().second] = st.top().second - i - 1;
            st.pop();
        }
        st.push({ heights[i], i });
    }
    for (int i = 0; i < n; i++) {
        if (foward[i] == -1) {
            foward[i] = n - i - 1;
        }
    }

    for (int i = 0; i < n; i++) {
        if (backwards[i] == -1) {
            backwards[i] = i;
        }
    }
    int ans = 0;
    for (int i = 0; i < n; i++) {
        int width = backwards[i] + 1 + foward[i];
        ans = max(ans, width * heights[i]);
    }

    return ans;
}
int main() {

    ios::sync_with_stdio(0);
    cin.tie(0);

    return 0;
}

File: python/stack/largestRectangle.py
import sys
    import threading
    from typing import List, Tuple, Dict, Set
from collections import defaultdict, deque


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        foward = [-1] * len(heights)
        backwards = [-1] * len(heights)
        stack = []

        for i, h in enumerate(heights):
            while stack and h < stack[-1][0]:
                foward[stack[-1][1]] = i - stack[-1][1] - 1
                stack.pop()
            stack.append((h, i))

        stack.clear()

        for i in range(len(heights) - 1, -1, -1):
            h = heights[i]
            while stack and h < stack[-1][0]:
                backwards[stack[-1][1]] = stack[-1][1] - i - 1
                stack.pop()
            stack.append((h, i))

        for i in range(len(foward)):
            if foward[i] == -1:
                foward[i] = len(foward) - i - 1

        for i in range(len(backwards)):
            if backwards[i] == -1:
                backwards[i] = i

        ans = 0
        for i in range(len(heights)):
            width = backwards[i] + 1 + foward[i]
            ans = max(ans, width * heights[i])

        return ans


def main():
    s = Solution()
    array = [7, 1, 7, 2, 2, 4]
    print(s.largestRectangleArea(array))


if __name__ == "__main__":
    threading.Thread(target=main).start()
